File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        if(asteroids == null || asteroids.length == 0){
            return new int[]{};
        }

        Stack<Integer> st = new Stack<>();
        for(int ast: asteroids){
            if(st.isEmpty() ||ast > 0){
                st.push(ast);
            } else {
                if(ast > 0){
                    st.push(ast);
                } else {
                    while(!st.isEmpty() && st.peek() > 0 && st.peek() < Math.abs(ast)){
                        st.pop();
                    }
                    if(st.isEmpty() || st.peek() < 0){
                        st.push(ast);
                    } else if(st.peek() == Math.abs(ast)){
                        st.pop();
                    }
                }
            }
        }

        int[] result = new int[st.size()];
        for (int i = st.size() - 1; i >= 0; i--) {
            result[i] = st.pop();
        }
        
        return result;
    }
}

File: 0735-asteroid-collision/0735-asteroid-collision.py
class Solution(object):
    def asteroidCollision(self, asteroids):
        """
        :type asteroids: List[int]
        :rtype: List[int]
        """
        stack = []
        
        for ast in asteroids:
            if not stack or ast > 0:
                stack.append(ast)
            else:
                while stack and stack[-1] > 0 and stack[-1] < abs(ast):
                    stack.pop()
                if not stack or stack[-1] < 0:
                    stack.append(ast)
                elif stack[-1] == abs(ast):
                    stack.pop()
        
        return stack
